src/custom/utils/metadata.ts

```diff
@@ -43,3 +51,18 @@ export function generateAppDataDoc(metadata: MetadataDoc = {}): AppDataDoc {
     },
   }
 }
+
+export async function uploadMetadataDocToIpfs(appDataDoc: AppDataDoc): Promise<string> {
+  try {
+    const { create } = await import('ipfs-http-client')
+    const client = create({ url: getIpfsUri() })
+    const doc = safeStringify.stableStringify(appDataDoc)
+    const { cid } = await client.add(doc)
+    await pinata.pinByHash(cid.toString())
```

Comment:
Maybe pinByHash can be another method exposed in `api/ipfs` that just delegate to `pinata.pinByHash`